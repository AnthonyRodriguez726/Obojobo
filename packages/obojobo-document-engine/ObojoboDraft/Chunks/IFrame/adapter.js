import Common from 'Common'

const setModelStateProp = Common.util.setModelStateProp

export default {
	construct(model, attrs) {
		const s = setModelStateProp.bind(this, model, attrs)

		s('type', 'media', p => (p.toLowerCase() === 'webpage' ? 'webpage' : 'media'))

		let defaultNewWindow
		let defaultBorder
		let defaultFit
		let defaultControls

		switch (model.modelState.type) {
			case 'webpage':
				defaultNewWindow = true
				defaultBorder = true
				defaultFit = 'scroll'
				defaultControls = ['zoom', 'reload', 'expand']
				break

			case 'media':
			default:
				defaultNewWindow = false
				defaultBorder = false
				defaultFit = 'scale'
				defaultControls = ['reload', 'expand']
				break
		}

		s('newWindow', defaultNewWindow, p => p === true)
		s('border', defaultBorder)
		s('fit', defaultFit, p => p.toLowerCase(), ['scroll', 'scale'])
		s('src', null)
		s('width', null, p => parseInt(p, 10) || null)
		s('height', null, p => parseInt(p, 10) || null)
		s('zoom', 1, p => parseFloat(p) || 1)
		s('newWindowSrc', null)
		s('autoload', false, p => p === true)
		s('title', null)
		s('controls', defaultControls, p => p.split(',').map(c => c.toLowerCase().replace(/ /g, '')))
		s('expandedSize', 'full', p => p.toLowerCase(), ['full', 'restricted'])
	},

	clone(model, clone) {
		clone.modelState.type = model.modelState.type
		clone.modelState.src = model.modelState.src
		clone.modelState.width = model.modelState.width
		clone.modelState.height = model.modelState.height
		clone.modelState.zoom = model.modelState.zoom
		clone.modelState.border = model.modelState.border
		clone.modelState.newWindow = model.modelState.newWindow
		clone.modelState.newWindowSrc = model.modelState.newWindowSrc
		clone.modelState.autoload = model.modelState.autoload
		clone.modelState.fit = model.modelState.fit
		clone.modelState.expandedSize = model.modelState.expandedSize
		clone.modelState.title = model.modelState.title
		clone.modelState.controls = model.modelState.controls
	},

	toJSON(model, json) {
		json.content.type = model.modelState.type
		json.content.src = model.modelState.src
		json.content.width = model.modelState.width
		json.content.height = model.modelState.height
		json.content.zoom = model.modelState.zoom
		json.content.border = model.modelState.border
		json.content.newWindow = model.modelState.newWindow
		json.content.newWindowSrc = model.modelState.newWindowSrc
		json.content.autoload = model.modelState.autoload
		json.content.fit = model.modelState.fit
		json.content.expandedSize = model.modelState.expandedSize
		json.content.title = model.modelState.title
		json.content.controls = model.modelState.controls
	},

	toText(model) {
		return model.modelState.src
	}
}
