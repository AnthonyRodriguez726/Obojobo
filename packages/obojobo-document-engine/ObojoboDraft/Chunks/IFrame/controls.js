import './controls.scss'

import React from 'react'
import parseURL from 'url-parse'

import Common from 'Common'
const isOrNot = Common.util.isOrNot

const controls = props => {
	const controlsOpts = props.controlsOptions
	const newWindowToolTip = parseURL(props.newWindowSrc).hostname

	return (
		<div
			className={
				'obojobo-draft--chunks--iframe--controls' +
				isOrNot(controlsOpts.zoom, 'zoomable') +
				isOrNot(controlsOpts.reload, 'reloadable')
			}
		>
			{controlsOpts.reload ? (
				<div className="control-button-container reload">
					<button className="reload-button" onClick={props.reload}>
						Reload
					</button>
					<span className="tool-tip">Reload</span>
				</div>
			) : null}
			{controlsOpts.newWindow ? (
				<div className="new-window-link">
					<a target="_blank" rel="noopener noreferrer" href={props.newWindowSrc}>
						View in a new window
					</a>

					<span className="tool-tip">{newWindowToolTip}</span>
				</div>
			) : null}
			{controlsOpts.zoom ? (
				<div className="size-controls">
					<div className="zoom-controls">
						{props.isZoomResettable ? (
							<div className="control-button-container zoom-reset">
								<button className="zoom-reset-button" onClick={props.zoomReset}>
									Reset zoom
								</button>
								<span className="tool-tip">Reset zoom</span>
							</div>
						) : null}
						<div className="control-button-container zoom-out">
							<button
								disabled={props.isZoomOutDisabled}
								className="zoom-out-button"
								onClick={props.zoomOut}
							>
								Zoom out
							</button>
							<span className="tool-tip">
								{props.isZoomOutDisabled ? "Whoa that's tiny! ðŸ˜²" : 'Zoom out'}
							</span>
						</div>
						<div className="control-button-container zoom-in">
							<button
								disabled={props.isZoomInDisabled}
								className="zoom-in-button"
								onClick={props.zoomIn}
							>
								Zoom in
							</button>
							<span className="tool-tip">Zoom in</span>
							<span className="tool-tip">
								{props.isZoomInDisabled ? 'At maximum zoom' : 'Zoom in'}
							</span>
						</div>
					</div>
				</div>
			) : null}
		</div>
	)
}

export default controls
