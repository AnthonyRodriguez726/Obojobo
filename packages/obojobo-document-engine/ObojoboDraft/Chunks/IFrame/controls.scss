@import '~styles/includes';

.obojobo-draft--chunks--iframe--controls {
	@mixin tool-tip() {
		pointer-events: none;
		position: absolute;
		border: 2px solid $color-bg;
		white-space: nowrap;
		font-size: 0.8em;
		color: $color-bg;
		background: $color-action;
		border-radius: 3em;
		line-height: 1em;
		padding: 0.75em 1.5em;
		font-weight: bold;
		font-size: 1em;
		z-index: $z-index-above-all;
		font-family: $font-default;
	}

	&.is-expandable.is-zoomable,
	&.is-unexpandable.is-zoomable {
		> .size-controls > .zoom-controls {
			&:after {
				content: ' ';
				right: 1.6em;
				top: 0.2em;
				width: 1px;
				display: block;
				height: 0.8em;
				position: absolute;
				background: $color-shadow;
				pointer-events: none;
			}
		}
	}

	.control-button-container {
		display: inline-block;
		font-size: 1.1em;
		height: 1em;

		> button {
			@include user-select(none);

			border: none;
			width: 1em;
			height: 1em;
			font-size: 1em;
			overflow: hidden;
			padding: 0;
			border-radius: $dimension-rounded-radius / 2;
			color: $color-transparent;
			background-color: $color-bg;
			background-position: center;
			background-repeat: no-repeat;
			cursor: pointer;
			vertical-align: top;
			font-family: $font-default;
			transition: color $duration-animation-default;
			word-wrap: initial;
			white-space: nowrap;

			&:focus {
				position: relative;
			}
		}

		&:hover > button {
			background-color: $color-action;
		}

		> .tool-tip {
			@include tool-tip();

			display: none;
			right: 0.5em;
			bottom: 3.2em;
			font-size: 0.65em;
		}

		&:hover {
			> .tool-tip {
				display: inline-block;
			}
		}
	}

	> .reload {
		position: absolute;
		left: 0;
		top: 0;

		> .reload-button {
			background-image: url('./reload.svg');
			background-size: 75%;
		}

		> .tool-tip {
			left: 0.5em;
			right: inherit;
			bottom: 2.55em;
		}

		&:hover > .reload-button {
			background-image: url('./reload-hover.svg');
		}
	}

	> .size-controls {
		position: absolute;
		right: 0;
		top: 0em;
	}

	> .size-controls > .zoom-controls {
		display: inline-block;
	}

	> .size-controls > .expand {
		margin-left: 1em;

		> .expand-button {
			background-image: url('./expand-button.svg');
			background-size: 75%;
		}

		&:hover > .expand-button {
			background-image: url('./expand-button-hover.svg');
		}
	}

	> .size-controls > .unexpand {
		margin-left: 1em;

		> .unexpand-button {
			background-image: url('./unexpand-button.svg');
			background-size: 75%;
		}
		&:hover > .unexpand-button {
			background-image: url('./unexpand-button-hover.svg');
		}
	}

	> .size-controls > .zoom-controls > .zoom-reset {
		> .zoom-reset-button {
			color: #b497d3;
			overflow: visible;
			font-size: 0.5em;
			width: inherit;
			padding: 0.7em 0.7em;
			padding-top: 0.9em;
			line-height: 0;
			height: 2em;
			margin-top: 0;
			vertical-align: top;
			border-radius: $dimension-rounded-radius;
		}
		&:hover > .zoom-reset-button {
			color: white;
		}
	}

	> .size-controls > .zoom-controls > .zoom-out {
		> .zoom-out-button {
			background-image: url('./zoom-out.svg');
			background-size: 50%;

			&:disabled {
				opacity: 0.4;
			}
		}

		&:hover > .zoom-out-button {
			background-image: url('./zoom-out-hover.svg');
		}
	}

	> .size-controls > .zoom-controls > .zoom-in {
		> .zoom-in-button {
			background-image: url('./zoom-in.svg');
			background-size: 50%;
		}

		&:hover > .zoom-in-button {
			background-image: url('./zoom-in-hover.svg');
		}
	}

	> .new-window-link {
		position: relative;
		display: inline-block;
		margin-top: 0;
		font-family: $font-default;
		font-size: 0.7em;
		opacity: 0.75;
		line-height: 1.5em;
		vertical-align: top;

		> a {
			display: inline-block;
			color: $color-action;
			cursor: pointer;
		}

		> .tool-tip {
			@include tool-tip();

			top: -3.8em;
			left: 50%;
			transform: translate(-50%);
			display: none;
		}

		&:hover {
			opacity: 1;

			> .tool-tip {
				display: block;
			}
		}
	}
}
