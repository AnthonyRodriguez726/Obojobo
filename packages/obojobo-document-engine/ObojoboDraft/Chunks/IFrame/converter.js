const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = node.data.get('content') || {}
	json.children = []

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type

	// Set up the defaults for content in order to migrate safely from older versions
	const contentType = node.content.type || 'media'
	const defaultContent = {
		type: contentType,
		border: contentType !== 'media',
		fit: contentType === 'media' ? 'scale' : 'scroll',
		initalZoom: 1,
		autoload: false,
		controls: contentType === 'media' ? 'reload' : 'zoom,reload,new-window'
	}

	const finalContent = Object.assign(defaultContent, node.content)

	json.data = { content: finalContent }

	return json
}

export default { slateToObo, oboToSlate }
