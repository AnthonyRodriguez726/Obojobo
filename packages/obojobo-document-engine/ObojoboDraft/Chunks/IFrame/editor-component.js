import React from 'react'
import Common from 'Common'

import IframeProperties from './iframe-properties-modal'

import Controls from './controls'

import './editor-component.scss'

const { ModalUtil } = Common.util
const { Button } = Common.components
const isOrNot = Common.util.isOrNot

class IFrame extends React.Component {
	constructor(props) {
		super(props)

		this.state = {
			isPreviewing: false
		}
	}

	showIFramePropertiesModal() {
		ModalUtil.show(
			<IframeProperties
				content={this.props.node.data.get('content')}
				onConfirm={this.changeProperties.bind(this)}
			/>
		)
	}

	changeProperties(content) {
		const editor = this.props.editor
		const change = editor.value.change()

		ModalUtil.hide()

		change.setNodeByKey(this.props.node.key, {
			data: {
				content
			}
		})
		editor.onChange(change)
	}

	togglePreviewing() {
		this.setState(state => ({
			isPreviewing: !state.isPreviewing
		}))
	}

	renderEditToolbar() {
		return (
			<div className="iframe-toolbar">
				<Button onClick={this.togglePreviewing.bind(this)}>Toggle Preview</Button>
				<Button className="properties-button" onClick={this.showIFramePropertiesModal.bind(this)}>
					IFrame Properties
				</Button>
			</div>
		)
	}

	renderIFramePreview() {
		const content = this.props.node.data.get('content')

		const iframeStyle = {
			height: 100 / content.initialZoom + '%',
			width: 100 / content.initialZoom + '%',
			transform: 'scale(' + content.initialZoom + ')'
		}

		return (
			<div className={'iframe-container'}>
				<iframe
					ref="iframe"
					src={this.props.node.data.get('content').src}
					is
					frameBorder="0"
					allow="geolocation; microphone; camera; midi; encrypted-media; vr"
					style={iframeStyle}
				/>
			</div>
		)
	}

	render() {
		const content = this.props.node.data.get('content')
		const controlList = content.controls.split(',')

		const previewStyle = {
			width: (content.width || '710') + 'px',
			height: (content.height || '500') + 'px'
		}

		const className =
			'obojobo-draft--chunks--iframe viewer pad is-previewing' +
			isOrNot(content.border, 'bordered') +
			isOrNot(this.state.isPreviewing, 'showing') +
			' is-controls-enabled' +
			' is-not-missing-src' +
			isOrNot(content.initialZoom > 1, 'scaled-up') +
			isOrNot(this.state.iframeIsSelected, 'selected')

		const controlsOpts = {
			reload: controlList.includes('reload'),
			newWindow: controlList.includes('new-window'),
			zoom: controlList.includes('zoom')
		}

		return (
			<div className={className}>
				<div className={'container'} style={previewStyle}>
					{this.renderEditToolbar()}
					{this.state.isPreviewing ? this.renderIFramePreview() : null}
					{this.state.isPreviewing ? (
						<Controls newWindowSrc={content.src} controlsOptions={controlsOpts} isEditor />
					) : null}
				</div>
			</div>
		)
	}
}

export default IFrame
