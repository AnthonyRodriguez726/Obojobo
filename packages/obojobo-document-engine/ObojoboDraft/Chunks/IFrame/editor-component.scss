@import '~styles/includes';

.obojobo-draft--chunks--iframe {
	position: relative;

	.obojobo-draft--components--button {
		display: block;
	}

	.iframe-toolbar {
		position: absolute;
		top: 50%;
		left: 0;
		width: 100%;
		transform: translate(0, -50%);
		z-index: 3;

		.obojobo-draft--components--button {
			margin-top: 0.2em;
			margin-bottom: 0.2em;

			button {
				border: 1px solid $color-bg;
			}
		}
	}
}
