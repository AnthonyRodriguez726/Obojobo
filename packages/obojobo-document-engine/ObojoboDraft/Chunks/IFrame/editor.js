import React from 'react'
import { Block } from 'slate'
import emptyNode from './empty-node.json'

const IFRAME_NODE = 'ObojoboDraft.Chunks.IFrame'

class Node extends React.Component {
	constructor(props) {
		super(props)
		this.state = this.props.node.data.get('content')
	}

	handleSourceChange(event) {
		const editor = this.props.editor
		const change = editor.value.change()

		change.setNodeByKey(this.props.node.key, {
			data: {
				content: {
					src: event.target.value
				}
			}
		})
		editor.onChange(change)
	}

	renderInput() {
		return (
			<div>
				<input
					value={this.props.node.data.get('content').src}
					onChange={event => this.handleSourceChange(event)}
					onClick={event => event.stopPropagation()}
				/>
			</div>
		)
	}

	renderFrame() {
		const { isFocused } = this.props

		const wrapperStyle = {
			position: 'relative',
			outline: isFocused ? '2px solid blue' : 'none'
		}

		const maskStyle = {
			display: isFocused ? 'none' : 'block',
			position: 'absolute',
			top: '0',
			left: '0',
			height: '100%',
			width: '100%',
			cursor: 'cell',
			zIndex: 1
		}

		return (
			<div className={'obojobo-draft--chunks--iframe viewer'} style={wrapperStyle}>
				<div style={maskStyle} />
				<iframe
					is
					src={this.props.node.data.get('content').src}
					frameBorder="0"
					allowFullScreen="true"
				/>
			</div>
		)
	}

	render() {
		return (
			<div className={'component'}>
				{this.renderFrame()}
				{this.props.isSelected ? this.renderInput() : null}
			</div>
		)
	}
}

const insertNode = change => {
	change
		.insertBlock(Block.fromJSON(emptyNode))
		.focus()
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = node.data.get('content') || {}
	json.children = []

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type
	json.data = { content: node.content }

	return json
}

const plugins = {
	renderNode(props) {
		switch (props.node.type) {
			case IFRAME_NODE:
				return <Node {...props} {...props.attributes} />
		}
	},
	schema: {
		blocks: {
			'ObojoboDraft.Chunks.IFrame': {
				isVoid: true
			}
		}
	}
}

const IFrame = {
	name: IFRAME_NODE,
	components: {
		Node
	},
	helpers: {
		insertNode,
		slateToObo,
		oboToSlate
	},
	json: {
		emptyNode
	},
	plugins
}

export default IFrame
