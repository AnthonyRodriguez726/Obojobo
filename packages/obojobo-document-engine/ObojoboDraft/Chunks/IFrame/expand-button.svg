<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 18.65 18.66">
	<defs>
		<style>.cls-1{fill:#b497d3;}</style>
	</defs>
	<title>expand-button</title>
	<polygon transform="translate(-1 1)" class="cls-1" points="12.5 7.42 15.19 4.73 18.66 8.2 18.66 0 10.46 0 13.92 3.47 11.23 6.16 12.5 7.42"/>
	<polygon transform="translate(1 -1)" class="cls-1" points="6.16 11.23 3.47 13.92 0 10.46 0 18.66 8.2 18.66 4.73 15.19 7.42 12.5 6.16 11.23"/>
</svg>