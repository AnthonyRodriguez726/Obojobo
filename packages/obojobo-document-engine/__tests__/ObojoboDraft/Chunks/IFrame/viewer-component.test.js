import React from 'react'
import renderer from 'react-test-renderer'

import IFrame from '../../../../ObojoboDraft/Chunks/IFrame/viewer-component'
import OboModel from '../../../../__mocks__/_obo-model-with-chunks'

describe('IFrame', () => {
	let model = OboModel.create({
		id: 'id',
		type: 'ObojoboDraft.Chunks.IFrame',
		content: {
			src: 'http://www.example.com'
		}
	})

	let moduleData = {
		focusState: {}
	}

	test('IFrame component', () => {
		const component = renderer.create(<IFrame model={model} moduleData={moduleData} />)
		let tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})
})
