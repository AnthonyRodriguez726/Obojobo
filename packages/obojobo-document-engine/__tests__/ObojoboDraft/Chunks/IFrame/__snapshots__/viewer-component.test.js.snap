// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IFrame IFrame component 1`] = `
<div
  className="component"
  data-focussed={false}
  data-id="id"
  data-obo-component={true}
  data-type="ObojoboDraft.Chunks.IFrame"
  id="obo-id"
>
  <div
    className="obojobo-draft--chunks--iframe viewer"
  >
    <iframe
      allowFullScreen="true"
      frameBorder="0"
      src="http://www.example.com"
    />
  </div>
</div>
`;
